### PR TITLE
Improve `sc-service` API

### PR DESCRIPTION
**Pull request by nazar-pc**
This improves `sc-service` API by not requiring the whole `&Configuration`, using specific configuration options instead. `RpcConfiguration` was also extracted from `Configuration` to group all RPC options together.

We don't use Substrate's CLI and would rather not use `Configuration` either, but some key public functions require it even though they ignored most of the fields anyway.

`RpcConfiguration` is very helpful not just for consolidation of the fields, but also to finally make RPC optional for our use case, while Substrate still runs RPC server on localhost even if listening address is explicitly set to `None`, which is annoying (and I suspect there is a reason for it, so didn't want to change the default just yet).

While this is a breaking change, most developers will not notice it if they use higher-level APIs.

Fixes https://github.com/paritytech/polkadot-sdk/issues/2897

### CODE REVIEW

**Comment by nazar-pc** [comment]
Would `ExecutorConfiguration` be a good name for use with `new_wasm_executor` that combines the following options?:
* default_heap_pages
* wasm_method
* max_runtime_instances
* runtime_cache_size

This is probably the last thing that I'd like to stop using `&Configuration` for now.

**Comment by nazar-pc** [comment]
Resolved merge conflicts after https://github.com/paritytech/polkadot-sdk/pull/4792 and would really appreciate if someone takes a closer look here

File: substrate/client/cli/src/lib.rs [hunk 1]
```diff
@@ -242,7 +243,18 @@ pub trait SubstrateCli: Sized {
...
+			logger_hook(logger_builder, &config)
+		})?;
+
+		if config.role.is_authority() && config.network.public_addresses.is_empty() {
+			warn!(
+				"WARNING: No public address specified, validator node may not be reachable.
+				Consider setting `--public-addr` to the public IP address of this node.
+				This will become a hard requirement in future versions."
+			);
+		}
```
Comment (bkchr) [inline substrate/client/cli/src/lib.rs:256]: This should be put into `network_params`.

File: prdoc/pr_5364.prdoc [hunk 2]
```diff
@@ -0,0 +1,27 @@
+title: Improve `sc-service` API
+
+doc:
+  - audience: Node Dev
+    description: |
+      This improves `sc-service` API by not requiring the whole `&Configuration`, using specific configuration options
+      instead. `RpcConfiguration` and `ExecutorConfiguration` were also extracted from `Configuration` to group all RPC
+      and executor options together.
+      If `sc-service` is used as a library with lower-level APIs, `Configuration` can now be avoided in most cases.
```
Comment (bkchr) [inline prdoc/pr_5364.prdoc:9]:
```suggestion
      If `sc-service` is used as a library with lower-level APIs, `Configuration` can now be avoided in most cases.
      
      This mainly impacts you on your node implementation. There you need to change
      ```
     with_execution_method(config.wasm_method)
     ```
     
     to
     
     ```
     with_execution_method(config.executor.wasm_method)
     ```

     There are similar changes required as well, but all are around the initialization of the wasm executor.
```

**Review by bkchr — APPROVED** [review]
Two nitpicks.

**Comment by nazar-pc** [comment]
Applied suggestions

File: substrate/client/cli/src/commands/run_cmd.rs [hunk 3]
```diff
@@ -345,7 +346,17 @@ impl CliConfiguration for RunCmd {
 	}
 
 	fn network_params(&self) -> Option<&NetworkParams> {
-		Some(&self.network_params)
+		let network_params = &self.network_params;
+		let is_authority = self.role(self.is_dev().ok()?).ok()?.is_authority();
+		if is_authority && network_params.public_addr.is_empty() {
+			warn!(
```
Comment (lexnv) [inline substrate/client/cli/src/commands/run_cmd.rs:352]: nit: Could you please double check we don't silence this error? (ie start a local node validator without this addr config)

We may initialize the logger after logging some messages, and a few components of the CLI use the `eprintln!()` macro for logging for that purpose

**Review by lexnv — APPROVED** [review]

File: substrate/client/cli/src/commands/run_cmd.rs (hunk 3 above)
Comment (nazar-pc) [reply substrate/client/cli/src/commands/run_cmd.rs]: Good catch, updated

File: substrate/client/service/src/config.rs [hunk 4]
```diff
@@ -323,3 +287,64 @@ impl From<PathBuf> for BasePath {
 		BasePath::new(path)
 	}
 }
+
+/// RPC configuration.
+#[derive(Debug)]
+pub struct RpcConfiguration {
```
Comment (niklasad1) [inline substrate/client/service/src/config.rs:293]: Nice refactor indeed, but with the introduced `addr: Vec<RpcEndpoint>` all the configurations/settings is already included in there and these other fields are not used by the server at all.

So, it should be possible to remove this struct an just use `Option<Vec<RpcEndpoint>` IIRC.

Comment (niklasad1) [reply substrate/client/service/src/config.rs:293]: Ok, I was wrong it can be none

**Comment by niklasad1** [comment]
> while Substrate still runs RPC server on localhost even if listening address is explicitly set to None, which is annoying (and I suspect there is a reason for it, so didn't want to change the default just yet).

I suppose you want to start your service without any rpc server then? Can you open an issue for it please?
 
I don't remember why we do it but that can changed if desired also with the latest refactoring it should be possible do `addr: Some(vec![])` as workaround but I think it will throw an error when no rpc listen addr is found and bail early with the error.

**Pull request merged**